main.tf

```
@@ -39,9 +39,20 @@ locals {
 
   # If provided, set Dataflow worker to new user-managed service account;
   # otherwise, use Compute Engine default service account
-  dataflow_worker_service_account = ((var.dataflow_worker_service_account != "")
-    ? google_service_account.dataflow_worker_service_account[0].email
-  : "${data.google_project.project.number}-compute@developer.gserviceaccount.com")
+  dataflow_worker_service_account = ((var.dataflow_worker_service_account != "") ?
+    ((var.use_externally_managed_dataflow_sa == false) ?
+      google_service_account.dataflow_worker_service_account[0].email :
+    var.dataflow_worker_service_account) :
+  "${data.google_project.project.number}-compute@developer.gserviceaccount.com")
+
+  # Grant Dataflow worker service account required IAM roles over external resources
+  # **unless** using an externally managed worker service account (Option 1 and 2):
+  grant_service_account_roles = var.use_externally_managed_dataflow_sa == true
```

@rarsan shouldn't it be `!=`?
```suggestion
  grant_service_account_roles = var.use_externally_managed_dataflow_sa != true
```

---

That's right :) should be `== false` which is option 1 (default) and 2.

---

One more thing that it is not protacting from invalid combination
```
use_externally_managed_dataflow_sa  = true
dataflow_worker_service_account = ""
```
This could be solved like this:
```suggestion
  grant_service_account_roles = (var.use_externally_managed_dataflow_sa != true && var.dataflow_worker_service_account != "")
```

---

Apart of that I tested all 3 variants and they looks good.

---

Can we add a variable validation? You're right about this invalid combo but there could be other things that might break, and ideally TF plan fails.

Is conditional validation using other variables now supported? Any other alternative to achieve that check?

---

As for now it is not possible to reference another variable in 'validation' block. So, it is not possible to fail on 'plan'. It is possible to add some validation to locals and fail on apply. I will check this out.

---

@rarsan I checked the issue, and want to confirm that there is no good options to solve it now. But here are some workarounds:
1. Use `precondition` for some resource. But this is available only from `1.2.0` and I prefer to avoid bumping version so high. That would limit if somebody running older Terraform.
2. Use this [workaround](https://github.com/hashicorp/terraform/issues/25609#issuecomment-1057614400), but that is quite dirty.

I didn't find any other options to do that.
Please let me know how do you want to proceed.

---

Thanks @Lakhtenkov-iv for looking at these options.
I agree with you. I'm ok with the workaround given that it achieves the desired effect (fail fast in case of invalid input combination), and it's an outlier scenario, i.e. obscure error message not expected to occur often. We can add a TODO to replace with a custom condition check like `precondition` whenever TF version is bumped to >= 1.2.0

---

Added, please review.

---

LGTM. Thanks. Resolving this.